File: libraryStudy/app/src/main/java/com/amazing/tengfei/librarystudy/touch/CustomViewGroup.java
package com.amazing.tengfei.librarystudy.touch;

import android.content.Context;
import android.util.AttributeSet;
import android.util.Log;
import android.view.MotionEvent;
import android.view.ViewGroup;

import javax.security.auth.login.LoginException;

/**
 * @author 飞一般的感觉
 * date 2020/5/19 11:23 AM
 * email arrayadapter.cn@gmail.com
 * description
 */
public class CustomViewGroup extends ViewGroup {


    private static final String TAG = "CustomViewGroup_TAG";

    public CustomViewGroup(Context context) {
        super(context);
    }

    public CustomViewGroup(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public CustomViewGroup(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    public CustomViewGroup(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
    }

    @Override
    protected void onLayout(boolean changed, int l, int t, int r, int b) {

    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        Log.e(TAG, "onTouchEvent: CustomViewGroup");
        return super.onTouchEvent(event);
    }

    @Override
    public boolean dispatchTouchEvent(MotionEvent ev) {
        return super.dispatchTouchEvent(ev);
    }

    @Override
    public boolean onInterceptTouchEvent(MotionEvent ev) {
        return true;
    }
}

File: MyApplication/app/src/main/java/com/example/myapplication/CustomViewGroup.kt
package com.example.myapplication

import android.content.Context
import android.util.AttributeSet
import android.util.Log
import android.view.MotionEvent
import android.widget.RelativeLayout

/**
 * @author tengfei
 * date 2020-03-25 10:36
 * email arrayadapter.cn@gmail.com
 * description
 */
class CustomViewGroup @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : RelativeLayout(context, attrs, defStyleAttr) {


    companion object {
        const val TAG = "CustomViewGroup_test"
    }


    override fun dispatchTouchEvent(ev: MotionEvent?): Boolean {
        Log.e(TAG, "CustomViewGroup  ${getAction(ev!!.action)}")
        return super.dispatchTouchEvent(ev)
    }

    var hasViewGroupIntercept: Boolean = false

    override fun onInterceptTouchEvent(ev: MotionEvent?): Boolean {
//        if (!hasViewGroupIntercept && ev?.action == MotionEvent.ACTION_MOVE) {
//            hasViewGroupIntercept = true
//            return true
//        }
        Log.e(TAG, "CustomViewGroup onInterceptTouchEvent")

        return true
    }

    override fun onTouchEvent(event: MotionEvent?): Boolean {
        Log.e(TAG, "CustomViewGroup onTouchEvent")
        return true
    }


    private fun getAction(action: Int?): String = when (action) {
        MotionEvent.ACTION_DOWN -> "dispatchTouchEvent ACTION_DOWN"
        MotionEvent.ACTION_MOVE -> "dispatchTouchEvent ACTION_MOVE"
        MotionEvent.ACTION_UP -> "dispatchTouchEvent ACTION_UP"
        MotionEvent.ACTION_CANCEL -> "dispatchTouchEvent ACTION_CANCEL"
        else -> "other"
    }

}